src/components/CollapsableText/CollapsableText.tsx

```diff
@@ -1,59 +1,73 @@
-import React, { useEffect, useState } from 'react';
-import Button, { ButtonProps } from '../Button/Button';
+/* eslint-disable react/no-unused-prop-types */
+import React, { useState, useRef } from 'react';
+import { useOnResizeRef } from '../../hooks/useOnResizeRef';
+import Button from '../Button/Button';
 
-const Toggle = ({ children, ...props }: ButtonProps) => (
-  <Button color="link" size="sm" className="p-0 m-0 ms-2" {...props}>
-    {children}
-  </Button>
-);
+function getEllipsisStyle(maxLines: number) {
+  maxLines = Math.max(maxLines, 1);
+  return {
+    display: '-webkit-box',
+    // max-height for browsers that don't support -webkit-line-clamp.
+    maxHeight: `calc(1.5em * ${maxLines})`,
+    overflow: 'hidden',
+    WebkitLineClamp: maxLines,
+    WebkitBoxOrient: 'vertical',
+  } as const;
+}
 
 export interface CollapsableTextProps {
-  children?: string;
+  children?: React.ReactNode;
   collapsed?: boolean;
   lessLabel?: React.ReactNode;
-  maxLength?: number;
   moreLabel?: React.ReactNode;
+  /**  @deprecated maxLength has no effect. Use maxLines instead */
+  maxLength?: number;
+  maxLines?: number;
 }
 
-const CollapsableText = ({
-  children = '',
-  collapsed: defaultCollapsed = true,
-  lessLabel = 'Show Less',
-  maxLength = 256,
-  moreLabel = 'Show More',
-}: CollapsableTextProps) => {
-  const [collapsed, setCollapsed] = useState(defaultCollapsed);
-  const toggle = () => setCollapsed(!collapsed);
+export default function CollapsableText({
+  children,
+  collapsed = CollapsableText.defaultProps.collapsed,
+  lessLabel = CollapsableText.defaultProps.lessLabel,
+  moreLabel = CollapsableText.defaultProps.moreLabel,
+  maxLines = CollapsableText.defaultProps.maxLines,
+}: CollapsableTextProps) {
+  const [isCollapsed, setIsCollapsed] = useState(collapsed);
+  const [hideToggle, setHideToggle] = useState(false);
 
-  useEffect(() => setCollapsed(defaultCollapsed), [defaultCollapsed]);
+  const containerCallbackRef = useOnResizeRef((container: Element) => {
+    // Hide toggle button if collpased size is greater or same as expanded size.
+    setHideToggle(isCollapsed && container.clientHeight >= container.scrollHeight);
+  });
 
-  if (children.length < maxLength) {
-    return children;
-  }
-  if (collapsed) {
-    return (
-      <span>
-        {children.substring(0, maxLength).trim()}&hellip;{' '}
-        <Toggle onClick={() => toggle()}>{moreLabel}</Toggle>
-      </span>
-    );
+  // If the collapsed prop changes, update the state.
+  // This is bad practice, we should instead have a defaultCollapsed prop and the collapsed prop
```

Comment:
I agree. Do you want to add a TODO to address this down the line or do you intend to refactor collapsed to be controlled via prop in this PR?

Reply:
I want to limit the impact of this PR. I'll add a TODO.

Reply:
Makes sense!